internal/api/event.go

```diff
@@ -0,0 +1,101 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/directives"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+type AppReference struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// NewPromotionEventAnnotations returns annotations for a Promotion related event.
+// It may skip some fields when error occurred during serialization, to record event with best-effort.
+func NewPromotionEventAnnotations(
+	ctx context.Context,
+	actor string,
+	p *kargoapi.Promotion,
+	f *kargoapi.Freight,
+) map[string]string {
+	logger := logging.LoggerFromContext(ctx)
+
+	annotations := map[string]string{
+		kargoapi.AnnotationKeyEventProject:             p.GetNamespace(),
+		kargoapi.AnnotationKeyEventPromotionName:       p.GetName(),
+		kargoapi.AnnotationKeyEventFreightName:         p.Spec.Freight,
+		kargoapi.AnnotationKeyEventStageName:           p.Spec.Stage,
+		kargoapi.AnnotationKeyEventPromotionCreateTime: p.GetCreationTimestamp().Format(time.RFC3339),
+	}
+
+	if actor != "" {
+		annotations[kargoapi.AnnotationKeyEventActor] = actor
+	}
+	// All Promotion-related events are emitted after the promotion was created.
+	// Therefore, if the promotion knows who triggered it, set them as an actor.
+	if promoteActor, ok := p.Annotations[kargoapi.AnnotationKeyCreateActor]; ok {
+		annotations[kargoapi.AnnotationKeyEventActor] = promoteActor
+	}
+
+	if f != nil {
+		annotations[kargoapi.AnnotationKeyEventFreightCreateTime] = f.CreationTimestamp.Format(time.RFC3339)
+		annotations[kargoapi.AnnotationKeyEventFreightAlias] = f.Alias
+		if len(f.Commits) > 0 {
+			data, err := json.Marshal(f.Commits)
+			if err != nil {
+				logger.Error(err, "marshal freight commits in JSON")
+			} else {
+				annotations[kargoapi.AnnotationKeyEventFreightCommits] = string(data)
+			}
+		}
+		if len(f.Images) > 0 {
+			data, err := json.Marshal(f.Images)
+			if err != nil {
+				logger.Error(err, "marshal freight images in JSON")
+			} else {
+				annotations[kargoapi.AnnotationKeyEventFreightImages] = string(data)
+			}
+		}
+		if len(f.Charts) > 0 {
+			data, err := json.Marshal(f.Charts)
+			if err != nil {
+				logger.Error(err, "marshal freight charts in JSON")
+			} else {
+				annotations[kargoapi.AnnotationKeyEventFreightCharts] = string(data)
+			}
+		}
+	}
+
+	var apps []AppReference
+	for _, step := range p.Spec.Steps {
+		if step.Uses != "argocd-update" || step.Config == nil {
+			continue
+		}
+		var cfg directives.ArgoCDUpdateConfig
```

Comment:
We're trying really hard to make built-in directives "not special." Ideally, the rest of Kargo would have as little awareness of their internals as it will eventually have of the internals of _third-party_ directives.

That being said, I understand this might seem like the only practical way to get the information you're trying to include in the events. This, unfortunately, wouldn't be the first spot where we compromised on this ideal because it seemed like the only way to get something done.

Ultimately, I think this ties into a _much broader_ question of how coupled do we want to be or _not_ want to be to Argo CD. Implementation details aside, just including this information in the Promotion event to begin with says something about Kargo's relationship to Argo CD being special in some way. I think this may be a question for @jessesuen.

Reply:
> We're trying really hard to make built-in directives "not special." Ideally, the rest of Kargo would have as little awareness of their internals as it will eventually have of the internals of third-party directives.

The intention of maintaining flexibility and avoiding any special logics for built-in and third-party directives makes sense to me. However I'm still a bit unclear about the design here. From my perspective, this little awareness is from code level instead of user level. Users still see the results the directives processed, but developers are limited to access the output of runners they have written.

> Ultimately, I think this ties into a much broader question of how coupled do we want to be or not want to be to Argo CD. Implementation details aside, just including this information in the Promotion event to begin with says something about Kargo's relationship to Argo CD being special in some way.

I would also like to know the idea behind https://github.com/akuity/kargo/pull/2545. As a Kargo newbie, when I saw the `argocd-update` step in`Promotion` resource, I would expect there are some ArgoCD integrations with Kargo. Otherwise it seems we might not need this directive from the lower-level implementations. Seems like this can also be related to how we can integrate with projects that manage workloads?

Reply:
@hiddeco I tried to move the entire `api/v1alpha1/event.go` file and kubernetes events related files to `internal/event` in https://github.com/akuity/kargo/pull/2808. However as you can see, this would require moving multiple files since some helpers rely on functions within that file. Moving just one function to internal/event doesn’t seem to make much sense. I’d also like to wait for consensus on the best location for this function.